storage/indexer/src/db_indexer.rs

```diff
@@ -0,0 +1,391 @@
+                .send(None)
+                .expect("Failed to send None to DBIndexer committer");
+            handle
+                .join()
+                .expect("DBIndexer committer thread fails to join");
+        }
+    }
+}
+
+impl DBIndexer {
+    pub fn new(
+        db: Arc<DB>,
+        db_reader: Arc<dyn DbReader>,
+        config: &InternalIndexerDBConfig,
+    ) -> Self {
+        let (sender, reciver) = mpsc::channel();
+
+        let db_clone = db.clone();
+        let committer_handle = thread::spawn(move || {
+            let committer = DBCommitter::new(db, reciver);
+            committer.run();
+        });
+
+        Self {
+            db: db_clone,
+            main_db_reader: db_reader,
+            config: *config,
+            sender,
+            committer_handle: Some(committer_handle),
+        }
+    }
+
+    pub fn get_persisted_version(&self) -> Result<Version> {
+        // read the latest key from the db
+        self.db
+            .get::<InternalIndexerMetadataSchema>(&MetadataKey::LatestVersion)?
+            .map_or(Ok(0), |metavalue| Ok(metavalue.expect_version()))
+    }
+
+    pub fn event_enabled(&self) -> bool {
+        self.config.enable_event
+    }
+
+    pub fn transaction_enabled(&self) -> bool {
+        self.config.enable_transaction
+    }
+
+    fn get_internal_db_iter(
+        &self,
+        start_version: Version,
+        num_transactions: u64,
+    ) -> Result<Box<dyn Iterator<Item = Result<(Transaction, Vec<ContractEvent>)>> + '_>> {
+        let txn_iter = self
+            .main_db_reader
+            .get_transaction_iterator(start_version, num_transactions)?;
+        let event_vec_iter = self
+            .main_db_reader
+            .get_events_iterator(start_version, num_transactions)?;
+        let zipped = txn_iter
+            .zip(event_vec_iter)
+            .map(|(txn_res, event_vec_res)| {
+                let txn = txn_res?;
+                let event_vec = event_vec_res?;
+                Ok((txn, event_vec))
+            });
+        Ok(Box::new(zipped)
+            as Box<
+                dyn Iterator<Item = Result<(Transaction, Vec<ContractEvent>)>> + '_,
+            >)
+    }
+
+    fn get_num_of_transactions(&self, version: Version) -> Result<u64> {
+        let highest_version = self.main_db_reader.get_synced_version()?;
+        let num_transaction = if (version + self.config.batch_size as u64) < highest_version {
+            self.config.batch_size as u64
+        } else {
+            highest_version - version
+        };
+        Ok(num_transaction)
+    }
+
+    pub fn process_a_batch(&self, start_version: Option<Version>) -> Result<Version> {
+        let mut version = start_version.unwrap_or(0);
+
+        let num_transactions = self.get_num_of_transactions(version)?;
+        let mut db_iter = self.get_internal_db_iter(version, num_transactions)?;
+        let batch = SchemaBatch::new();
+        db_iter.try_for_each(|res| {
```

Comment:
nit: I'm a bit paranoid, but I fear zipping iterators in general -- could you please assert the number of results equals `num_transactions`? 

```
let mut count = 0;
db_iter.try_for_each(|| {
    ...
    count += 1;
})?;
ensure!(count == num_transactions)
```